File: test/basic-config.c
#include <stdlib.h>
#include <stdint.h>
#include <string.h>
#include "basic-config.h"

version_info *make_version_info(int major, int minor, int patch) {
  version_info *v = malloc(sizeof(version_info));
  v->major = major;
  v->minor = minor;
  v->patch = patch;
  return v;
}

options *make_options(int c) {
  options *o = malloc(sizeof(options));
  o->size = c;
  o->version = make_version_info(c, c + 1, c + 2);
  return o;
}

int64_t addI64(int64_t a, int64_t b) {
  return a + b;
}

int *arrayStuff(int arr[3]) {
  for (int i = 0; i < 3; i++) {
    arr[i] += 1;
  }
  return arr;
}

char *stringStuff(char *str) {
  int len = strlen(str);
  for (int i = 0; i < len; i++) {
    if (str[i] >= 'a' && str[i] <= 'z') {
      str[i] = str[i] - ('a' - 'A');
    }
  }
  return str;
}

void callback(void (*f)(int)) {
  f(777);
}

int plusOne(int x) {
  return x + 1;
}

int (*uncallback())(int) {
  return &plusOne;
}

File: test/basic-config.mjs
import * as basic from './build/basic-bindings.mjs'

function version2string(version) {
  return `{ major: ${version.major}, minor: ${version.minor}, patch: ${version.patch} }`
}

function options2string(options) {
  return `{ size: ${options.size}, version: ${version2string(options.version.deref())} }`
}

let version = basic.make_version_info(1, 2, 3).deref()
console.log(version2string(version))

version.major = 3
version.minor = 2
version.patch = 1
console.log(version2string(version))
version.destroy()

let options = basic.make_options(5).deref()
console.log(options2string(options))
options.destroy()

console.log(basic.addI64(5n, 6n))

let array = [1, 2, 3]
let result = basic.arrayStuff(array)
let newArray = []
for (let i = 0; i < 3; i++) {
  newArray.push(result.offset(i).deref())
}
console.log(newArray)

let string = "hello"
console.log(basic.stringStuff(string))

basic.callback(x => console.log(x))

console.log(basic.uncallback()(1));
